File: packages/arcis-python/arcis/sanitizers/xxe.py
"""
Arcis XXE (XML External Entity) injection prevention.

Detects and sanitizes XXE payloads including DOCTYPE declarations,
ENTITY definitions, SYSTEM/PUBLIC references, parameter entities,
and CDATA abuse.
"""

import re

# Detection patterns — used by detect_xxe()
_XXE_DETECT_PATTERNS = [
    # DOCTYPE declaration
    re.compile(r"<!DOCTYPE\b", re.IGNORECASE),
    # ENTITY declaration
    re.compile(r"<!ENTITY\b", re.IGNORECASE),
    # SYSTEM keyword with URI
    re.compile(r"\bSYSTEM\s+[\"']", re.IGNORECASE),
    # PUBLIC keyword with URI
    re.compile(r"\bPUBLIC\s+[\"']", re.IGNORECASE),
    # Parameter entity reference (%entity;)
    re.compile(r"%\s*\w+\s*;"),
    # CDATA section
    re.compile(r"<!\[CDATA\[", re.IGNORECASE),
]

# Removal patterns — used by sanitize_xxe()
_XXE_REMOVE_PATTERNS = [
    # Full DOCTYPE block with optional internal subset: <!DOCTYPE ... [...]>
    re.compile(r"<!DOCTYPE\s[^[>]*(?:\[[^\]]*\]\s*)?>|<!DOCTYPE\s[^>]*>", re.IGNORECASE),
    # Full ENTITY declaration: <!ENTITY ... >
    re.compile(r"<!ENTITY[^>]*>", re.IGNORECASE),
    # CDATA sections: <![CDATA[ ... ]]>
    re.compile(r"<!\[CDATA\[[\s\S]*?\]\]>", re.IGNORECASE),
]


def detect_xxe(value: str) -> bool:
    """
    Check if a string contains XXE patterns.

    Returns True if any XXE injection pattern is detected.
    Does not modify the input — use sanitize_xxe() for that.
    """
    if not isinstance(value, str):
        return False

    for pattern in _XXE_DETECT_PATTERNS:
        if pattern.search(value):
            return True

    return False


def sanitize_xxe(value: str) -> str:
    """
    Sanitize a string by removing XXE payloads.

    Strips DOCTYPE declarations, ENTITY definitions, and CDATA sections.
    """
    if not isinstance(value, str):
        raise TypeError(f"sanitize_xxe expects str, got {type(value).__name__}")

    result = value
    for pattern in _XXE_REMOVE_PATTERNS:
        result = pattern.sub("", result)

    return result

File: packages/arcis-node/src/sanitizers/xxe.ts
/**
 * @module @arcis/node/sanitizers/xxe
 * XML External Entity (XXE) injection prevention
 */

import type { SanitizeResult, ThreatInfo } from '../core/types';

/**
 * XXE detection patterns (ReDoS-safe).
 *
 * Covers DOCTYPE declarations, ENTITY definitions, SYSTEM/PUBLIC references,
 * parameter entities, and CDATA abuse.
 */
/**
 * Billion-laughs defense: cap raw XML input length and the count of
 * entity references. A valid document rarely needs more than a handful of
 * entities; thousands of `&foo;` references is the classic bomb shape.
 */
const MAX_XXE_INPUT_BYTES = 1_000_000; // 1 MB — above any reasonable config/SOAP payload
const MAX_ENTITY_REFERENCES = 64;

const XXE_DETECT_PATTERNS = [
  /** DOCTYPE declaration */
  /<!DOCTYPE\b/gi,
  /** ENTITY declaration */
  /<!ENTITY\b/gi,
  /** SYSTEM keyword with URI */
  /\bSYSTEM\s+["']/gi,
  /** PUBLIC keyword with URI */
  /\bPUBLIC\s+["']/gi,
  /** Parameter entity reference (%entity;) */
  /%\s*\w+\s*;/g,
  /** CDATA section (often used to smuggle payloads) */
  /<!\[CDATA\[/gi,
] as const;

/** Removal patterns — strip the dangerous XML constructs */
const XXE_REMOVE_PATTERNS = [
  /** Full DOCTYPE block with optional internal subset: <!DOCTYPE ... [...]> */
  /<!DOCTYPE\s[^[>]*(?:\[[^\]]*\]\s*)?>|<!DOCTYPE\s[^>]*>/gi,
  /** Full ENTITY declaration: <!ENTITY ... > */
  /<!ENTITY[^>]*>/gi,
  /** CDATA sections: <![CDATA[ ... ]]> */
  /<!\[CDATA\[[\s\S]*?\]\]>/gi,
] as const;

/**
 * Sanitizes a string to prevent XXE attacks.
 * Removes DOCTYPE, ENTITY, and CDATA constructs.
 */
export function sanitizeXxe(input: string, collectThreats?: false): string;
export function sanitizeXxe(input: string, collectThreats: true): SanitizeResult;
export function sanitizeXxe(input: string, collectThreats = false): string | SanitizeResult {
  if (typeof input !== 'string') {
    return collectThreats
      ? { value: String(input), wasSanitized: false, threats: [] }
      : String(input);
  }

  const threats: ThreatInfo[] = [];
  let value = input;
  let wasSanitized = false;

  // Billion-laughs defense: oversize input or many entity refs → flatten to empty.
  // Safer to discard than to attempt partial sanitization of a bomb payload.
  if (value.length > MAX_XXE_INPUT_BYTES) {
    if (collectThreats) {
      threats.push({ type: 'xxe', pattern: 'oversize_input', original: `length=${value.length}` });
    }
    return collectThreats ? { value: '', wasSanitized: true, threats } : '';
  }
  const entityRefs = value.match(/&\w+;/g);
  if (entityRefs && entityRefs.length > MAX_ENTITY_REFERENCES) {
    if (collectThreats) {
      threats.push({ type: 'xxe', pattern: 'entity_expansion', original: `count=${entityRefs.length}` });
    }
    return collectThreats ? { value: '', wasSanitized: true, threats } : '';
  }

  for (const pattern of XXE_REMOVE_PATTERNS) {
    pattern.lastIndex = 0;
    if (pattern.test(value)) {
      pattern.lastIndex = 0;

      if (collectThreats) {
        const matches = value.match(pattern);
        if (matches) {
          for (const match of matches) {
            threats.push({
              type: 'xxe',
              pattern: pattern.source,
              original: match,
            });
          }
        }
      }

      value = value.replace(pattern, '');
      wasSanitized = true;
    }
  }

  if (collectThreats) {
    return { value, wasSanitized, threats };
  }

  return value;
}

/**
 * Checks if a string contains XXE patterns.
 * Does not sanitize — use sanitizeXxe() for that.
 *
 * @param input - The string to check
 * @returns True if XXE patterns detected
 */
export function detectXxe(input: string): boolean {
  if (typeof input !== 'string') return false;

  for (const pattern of XXE_DETECT_PATTERNS) {
    pattern.lastIndex = 0;
    if (pattern.test(input)) {
      return true;
    }
  }

  return false;
}
